File: Data Structures & Algorithms/surrounded-regions/submission-0.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        n,m = len(board), len(board[0])
        q = []
        vis = [[False for _ in range(m)] for _ in range(n)]
        for i in range(n):
            if board[i][0] == 'O' and not vis[i][0]:
                q.append((i,0))
                vis[i][0] = True
            if board[i][m-1] == 'O' and not vis[i][m-1]:
                q.append((i,m-1))
                vis[i][m-1] = True

        for i in range(m):
            if board[0][i] == 'O' and not vis[0][i]:
                q.append((0,i))
                vis[0][i] = True
            if board[n-1][i] == 'O' and not vis[n-1][i]:
                q.append((n-1,i))
                vis[n-1][i] = True
        
        delr = [-1, 0, 1, 0]
        delc = [0, -1, 0, 1]

        while q:
            x,y = q.pop(0)
            for i in range(4):
                nrow, ncol = x + delr[i], y + delc[i]
                if 0 <= nrow < n and 0 <= ncol < m and board[nrow][ncol] == 'O' and not vis[nrow][ncol]:
                    vis[nrow][ncol] = True
                    q.append((nrow, ncol))

        for i in range(n):
            for j in range(m):
                if board[i][j] == 'O' and not vis[i][j]:
                    board[i][j] = 'X'
            
        

File: Data Structures & Algorithms/search-for-word-ii/submission-0.cpp
class TrieNode{
public:
    TrieNode* child[26] = {NULL};
    bool isEnd = false;
};

class Solution {
public:
    void  addWord(TrieNode* root, string word){
        if(!root) root = new TrieNode();
        TrieNode* curr = root;
        for(auto c: word){
            if(!curr->child[c-'a']){
                curr->child[c-'a'] = new TrieNode();
            }
            curr = curr->child[c-'a'];
        }
        curr->isEnd = true;
    }
    int delR[4] = {-1, 0, 1, 0};
    int delC[4] = {0, -1, 0, 1}; 
    void search(vector<vector<char>>& board, vector<string>& ans, int row, int col, TrieNode* curr, string temp){
        if(!curr) return;
        if(curr->isEnd){
            ans.push_back(temp);
            curr->isEnd = false;
        }
        char og = board[row][col];
        board[row][col] = '-';
        for(int i = 0; i < 4; i++){
            int nrow = row + delR[i];
            int ncol = col + delC[i];
            if(nrow >= 0 && ncol >= 0&& nrow < board.size() && ncol < board[0].size() && board[nrow][ncol] != '-' && curr->child[board[nrow][ncol]-'a']){
                search(board, ans, nrow, ncol, curr->child[board[nrow][ncol] - 'a'], temp + board[nrow][ncol]);                
            }
        }
        board[row][col] = og;
    }

    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        vector<string> ans;
        TrieNode* root = new TrieNode();
        for(auto word: words){
            addWord(root, word);
        }
        int n = board.size(), m = board[0].size();
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                char ch = board[i][j];
                if(root->child[ch-'a']) search(board, ans, i, j, root->child[ch-'a'], string(1,ch));
            }
        }
        return ans;
    }
};
